### PR TITLE
feat: Remove Unpin requirement from Framed[Read,Write,]

### PR DESCRIPTION
**Pull request by Marwes**
cc #1252

### CODE REVIEW

File: tokio-util/src/codec/framed.rs [hunk 1]
```diff
@@ -17,11 +18,38 @@ use std::task::{Context, Poll};
...
 
-pub(crate) struct Fuse<T, U>(pub(crate) T, pub(crate) U);
+#[pin_project]
+pub(crate) struct Fuse<T, U> {
+    #[pin]
+    pub(crate) io: T,
+    pub(crate) codec: U,
+}
+
+pub(crate) trait ProjectFuse {
```
Comment (taiki-e) [inline tokio-util/src/codec/framed.rs]: Why this trait is needed? 
(FYI: [the projection type generated by pin-project has basically the same visibility as the original type.](https://github.com/taiki-e/pin-project/blob/c75d010eed7bf8146dd2f5228ed66c3cd06d50e9/pin-project-internal/src/lib.rs#L63-L65))

Comment (Marwes) [reply tokio-util/src/codec/framed.rs]: It abstracts over `FramedRead2` being either `FramedRead2<FramedWrite2<Fuse<T, U>>>` or `FramedRead2<Fuse<T, U>>` and lets the `io` and `codec` parts be extracted either way.

Without it `encode/decode` can't be called due to `Self` being pinned.

Comment (taiki-e) [reply tokio-util/src/codec/framed.rs]: Oh, that makes sense. Thanks for clarifying.

Comment (hawkw) [reply tokio-util/src/codec/framed.rs]: I think it could be helpful to add a comment here explaining that?

**Pull request merged**